File: docs/edgedepth_original/src/core/tpo_manager.cpp
#include "tpo_manager.h"
#include <cstdio>
#include <numeric>
#include <functional>
#include <unordered_set>

// ═══════════════════════════════════════════════════════════════════════════════
// Session boundary helpers
// ═══════════════════════════════════════════════════════════════════════════════

// Align a timestamp (ms) down to session start boundary.
// For daily sessions: floor to 00:00 UTC of that day.
static int64_t align_session_start(int64_t ts_ms, int session_hours) {
    const int64_t period_ms = static_cast<int64_t>(session_hours) * 3600LL * 1000LL;
    // Floor to midnight UTC for daily (works for 24h multiples too)
    if (session_hours == 24) {
        const int64_t day_ms = 86400LL * 1000LL;
        return (ts_ms / day_ms) * day_ms;
    }
    // For other periods, align to epoch
    return (ts_ms / period_ms) * period_ms;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Build sessions from candle data
// ═══════════════════════════════════════════════════════════════════════════════

void TPOManager::build_sessions(
    const std::string& symbol,
    const double* timestamps,
    const double* highs,
    const double* lows,
    size_t candle_count,
    int64_t timeframe_sec,
    double tick_per_row)
{
    if (!timestamps || !highs || !lows || candle_count == 0) return;

    // Deliberately unused. A TPO period is 30 minutes whatever the chart is
    // showing - that is what makes two traders' profiles of the same session
    // comparable - so the caller's timeframe does not size the periods. The
    // parameter stays in the signature because the caller has it to hand and a
    // future non-30m profile grain would need it.
    (void)timeframe_sec;

    // Debounce. The chart calls this on every frame it scrolls, and rebuilding a
    // whole profile per frame is wasted work, so a build is skipped when nothing
    // it depends on has moved.
    //
    // "Depends on" is the candle window AND the config that shapes the
    // computation. Leaving the config out made the settings UI look dead:
    // changing Value Area % or the session period recomputed nothing until the
    // candles happened to move, so a static chart kept showing a profile built
    // at the old setting.
    //
    // Display-only settings are deliberately NOT in the key. They are read at
    // render time, so hashing them would rebuild for nothing, and a rebuild
    // discards each session's expanded flag. ticks_per_row_setting is absent for
    // the opposite reason: the caller already resolves it into tick_per_row.
    uint64_t key = 1469598103934665603ULL;          // FNV-1a 64 offset basis
    const auto mix = [&key](uint64_t value) {
        key = (key ^ value) * 1099511628211ULL;     // FNV-1a 64 prime
    };
    mix(static_cast<uint64_t>(candle_count));
    mix(static_cast<uint64_t>(static_cast<int64_t>(timestamps[0])));
    mix(static_cast<uint64_t>(static_cast<int64_t>(timestamps[candle_count - 1])));
    mix(static_cast<uint64_t>(std::llround(tick_per_row * 1000.0)));
    mix(static_cast<uint64_t>(session_period_hours));
    mix(static_cast<uint64_t>(std::llround(value_area_pct * 1000.0)));

    auto& prev_key = last_build_hash_[symbol];
    if (key == prev_key) return;
    prev_key = key;

    const int64_t session_ms = static_cast<int64_t>(session_period_hours) * 3600LL * 1000LL;
    auto& sessions = data_[symbol];
    sessions.clear();

    // Group candles into sessions
    size_t i = 0;
    while (i < candle_count) {
        int64_t candle_ts = static_cast<int64_t>(timestamps[i]);
        int64_t sess_start = align_session_start(candle_ts, session_period_hours);
        int64_t sess_end = sess_start + session_ms;

        // Find all candles in this session
        size_t sess_first = i;
        while (i < candle_count &&
               static_cast<int64_t>(timestamps[i]) < sess_end) {
            ++i;
        }
        size_t sess_last = i; // exclusive

        if (sess_last <= sess_first) continue;

        TPOSession session;
        build_single_session(session, timestamps, highs, lows,
                             sess_first, sess_last,
                             sess_start, sess_end, tick_per_row);
        sessions.push_back(std::move(session));
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Build a single session from candles in range [start_idx, end_idx)
// ═══════════════════════════════════════════════════════════════════════════════

void TPOManager::build_single_session(
    TPOSession& session,
    const double* timestamps,
    const double* highs,
    const double* lows,
    size_t start_idx, size_t end_idx,
    int64_t session_start_ms,
    int64_t session_end_ms,
    double tick_per_row)
{
    session.session_start_ms = session_start_ms;
    session.session_end_ms = session_end_ms;

    // Find session price range
    double sess_high = -1e18, sess_low = 1e18;
    for (size_t i = start_idx; i < end_idx; ++i) {
        if (highs[i] > sess_high) sess_high = highs[i];
        if (lows[i] < sess_low)   sess_low = lows[i];
    }
    session.session_high = sess_high;
    session.session_low = sess_low;

    // Auto tick_per_row if not specified
    double tpr = tick_per_row;
    if (tpr <= 0.0) {
        double range = sess_high - sess_low;
        if (range <= 0.0) range = 1.0;
        // Target ~40-60 rows per session
        tpr = range / 50.0;
        if (tpr < 1e-8) tpr = 1e-8;
    }
    session.tick_per_row = tpr;

    // Compute row grid
    int row_lo_idx = static_cast<int>(std::floor(sess_low / tpr));
    int row_hi_idx = static_cast<int>(std::floor(sess_high / tpr));
    int num_rows = row_hi_idx - row_lo_idx + 1;
    if (num_rows <= 0 || num_rows > 10000) return; // sanity

    session.rows.resize(num_rows);
    for (int r = 0; r < num_rows; ++r) {
        session.rows[r].row_idx = r;
        session.rows[r].price_lo = (row_lo_idx + r) * tpr;
        session.rows[r].price_hi = (row_lo_idx + r + 1) * tpr;
    }

    // Assign blocks: group candles into 30m periods regardless of actual timeframe.
    // Multiple candles within the same 30m window contribute to the same period
    // but we only record each period once per row (dedup via period_idx tracking).
    static constexpr int64_t PERIOD_MS = 30LL * 60 * 1000; // 30 minutes
    session.total_periods = 0;
    session.total_blocks = 0;

    // Track which (row, period) pairs have been assigned to avoid duplicates
    // when multiple candles fall within the same 30m period
    // Use a flat set: row_idx * 1000 + period_idx (safe for < 1000 periods/day)
    std::unordered_set<int64_t> assigned;

    for (size_t i = start_idx; i < end_idx; ++i) {
        int64_t candle_ts = static_cast<int64_t>(timestamps[i]);
        // Period index = which 30m slot within the session
        int period_idx = static_cast<int>((candle_ts - session_start_ms) / PERIOD_MS);
        session.total_periods = std::max(session.total_periods, period_idx + 1);

        // Which rows does this candle touch?
        int candle_row_lo = static_cast<int>(std::floor(lows[i] / tpr)) - row_lo_idx;
        int candle_row_hi = static_cast<int>(std::floor(highs[i] / tpr)) - row_lo_idx;
        candle_row_lo = std::max(0, candle_row_lo);
        candle_row_hi = std::min(num_rows - 1, candle_row_hi);

        for (int r = candle_row_lo; r <= candle_row_hi; ++r) {
            int64_t key = static_cast<int64_t>(r) * 1000 + period_idx;
            if (assigned.insert(key).second) {
                // First time this (row, period) pair seen
                session.rows[r].blocks.push_back({period_idx});
                session.total_blocks++;
            }
        }
    }

    // Cache max block count
    session.max_block_count = 0;
    for (const auto& row : session.rows) {
        int bc = static_cast<int>(row.blocks.size());
        if (bc > session.max_block_count) session.max_block_count = bc;
    }

    // Compute derived data
    compute_poc(session);
    compute_value_area(session);
    detect_single_prints(session);
    detect_poor_high_low(session);
    compute_initial_balance(session);
}

// ═══════════════════════════════════════════════════════════════════════════════
// POC - widest row. Tie-break: closest to session midpoint.
// ═══════════════════════════════════════════════════════════════════════════════

void TPOManager::compute_poc(TPOSession& session) {
    if (session.rows.empty()) return;

    double midpoint = (session.session_high + session.session_low) * 0.5;
    int best_idx = 0;
    int best_count = 0;
    double best_dist = 1e18;

    for (size_t i = 0; i < session.rows.size(); ++i) {
        int count = static_cast<int>(session.rows[i].blocks.size());
        double row_mid = (session.rows[i].price_lo + session.rows[i].price_hi) * 0.5;
        double dist = std::abs(row_mid - midpoint);

        if (count > best_count || (count == best_count && dist < best_dist)) {
            best_idx = static_cast<int>(i);
            best_count = count;
            best_dist = dist;
        }
    }

    session.poc_row_idx = best_idx;
    session.poc_price = (session.rows[best_idx].price_lo +
                         session.rows[best_idx].price_hi) * 0.5;
    session.rows[best_idx].is_poc = true;
}

// ═══════════════════════════════════════════════════════════════════════════════
// Value Area - expand from POC until value_area_pct of total blocks captured.
// Compare row above vs below, take whichever has more blocks.
// ═══════════════════════════════════════════════════════════════════════════════

void TPOManager::compute_value_area(TPOSession& session) {
    if (session.rows.empty() || session.total_blocks == 0) return;

    int target_blocks = static_cast<int>(
        std::ceil(session.total_blocks * value_area_pct));

    int poc_idx = session.poc_row_idx;
    int num_rows = static_cast<int>(session.rows.size());

    // Start with POC row
    int accumulated = static_cast<int>(session.rows[poc_idx].blocks.size());
    session.rows[poc_idx].is_value_area = true;

    int upper = poc_idx + 1; // next row above to consider
    int lower = poc_idx - 1; // next row below to consider

    while (accumulated < target_blocks && (upper < num_rows || lower >= 0)) {
        int above_count = (upper < num_rows)
            ? static_cast<int>(session.rows[upper].blocks.size()) : -1;
        int below_count = (lower >= 0)
            ? static_cast<int>(session.rows[lower].blocks.size()) : -1;

        if (above_count >= below_count && above_count >= 0) {
            session.rows[upper].is_value_area = true;
            accumulated += above_count;
            upper++;
        } else if (below_count >= 0) {
            session.rows[lower].is_value_area = true;
            accumulated += below_count;
            lower--;
        } else {
            break;
        }
    }

    // VAH = top of highest VA row, VAL = bottom of lowest VA row
    session.vah = session.session_low;
    session.val = session.session_high;
    for (const auto& row : session.rows) {
        if (row.is_value_area) {
            if (row.price_hi > session.vah) session.vah = row.price_hi;
            if (row.price_lo < session.val) session.val = row.price_lo;
        }
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Single Prints - rows with exactly 1 block, where row above AND below
// each have >1 block. Top and bottom rows are excluded.
// ═══════════════════════════════════════════════════════════════════════════════

void TPOManager::detect_single_prints(TPOSession& session) {
    int n = static_cast<int>(session.rows.size());
    if (n < 3) return;

    for (int i = 1; i < n - 1; ++i) {
        if (session.rows[i].blocks.size() == 1 &&
            session.rows[i - 1].blocks.size() > 1 &&
            session.rows[i + 1].blocks.size() > 1) {
            session.rows[i].is_single_print = true;
        }
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Poor High/Low - top/bottom row has ≥2 blocks (price stayed there)
// ═══════════════════════════════════════════════════════════════════════════════

void TPOManager::detect_poor_high_low(TPOSession& session) {
    if (session.rows.empty()) return;
    session.has_poor_high = (session.rows.back().blocks.size() >= 2);
    session.has_poor_low = (session.rows.front().blocks.size() >= 2);
}

// ═══════════════════════════════════════════════════════════════════════════════
// Initial Balance - range of first 2 periods (00:00-01:00 UTC)
// ═══════════════════════════════════════════════════════════════════════════════

void TPOManager::compute_initial_balance(TPOSession& session)
{
    session.ib_high = -1e18;
    session.ib_low = 1e18;
    bool found = false;

    // The IB range comes off the ROW GRID, not off the raw candles: it is the
    // span of the rows that periods 0 and 1 printed into, so it lines up with
    // the profile the user is looking at rather than with the underlying wicks.
    for (auto& row : session.rows) {
        for (const auto& blk : row.blocks) {
            if (blk.period_idx < 2) {
                row.is_initial_balance = true;
                if (row.price_hi > session.ib_high) session.ib_high = row.price_hi;
                if (row.price_lo < session.ib_low)  session.ib_low = row.price_lo;
                found = true;
                break;
            }
        }
    }

    if (!found) {
        session.ib_high = 0.0;
        session.ib_low = 0.0;
    }
}

// ═══════════════════════════════════════════════════════════════════════════════
// Public accessors
// ═══════════════════════════════════════════════════════════════════════════════

const std::vector<TPOSession>*
TPOManager::get_sessions(const std::string& symbol) const {
    auto it = data_.find(symbol);
    if (it == data_.end() || it->second.empty()) return nullptr;
    return &it->second;
}

bool TPOManager::has_data(const std::string& symbol) const {
    auto it = data_.find(symbol);
    return it != data_.end() && !it->second.empty();
}

void TPOManager::toggle_expand(const std::string& symbol, size_t session_idx) {
    auto it = data_.find(symbol);
    if (it == data_.end()) return;
    if (session_idx < it->second.size()) {
        it->second[session_idx].expanded = !it->second[session_idx].expanded;
    }
}

void TPOManager::clear(const std::string& symbol) {
    data_.erase(symbol);
    last_build_hash_.erase(symbol);
}

File: frontend/src/lib/edgedepth/core/tpo_manager.ts
// tpo_manager.h/.cpp — exact port line by line, space by space, bracket by bracket, as is
// Original: https://github.com/edgedepthhq/edgedepth-terminal.git src/core/tpo_manager.h/.cpp
// TPO Market Profile from 30m candles, sessions 00:00 UTC, POC VAH VAL single prints poor high/low IB

export type TPOBlock = { period_idx:number; };
export type TPORow = { price_lo:number; price_hi:number; row_idx:number; blocks:TPOBlock[]; block_count:number; is_poc:boolean; is_value_area:boolean; is_single_print:boolean; is_initial_balance:boolean; };
export type TPOSession = {
  session_start_ms:number; session_end_ms:number; rows:TPORow[]; total_periods:number; total_blocks:number; max_block_count:number;
  poc_price:number; poc_row_idx:number; vah:number; val:number; session_high:number; session_low:number; ib_high:number; ib_low:number;
  has_poor_high:boolean; has_poor_low:boolean; tick_per_row:number; expanded:boolean;
};

export class TPOManager {
  private data = new Map<string, TPOSession[]>();
  private last_build_hash = new Map<string, number>();
  session_period_hours=24;
  ticks_per_row_setting=0;
  value_area_pct=0.70;
  show_poc_ray=true;
  show_vah_val_rays=true;
  show_single_prints=true;
  show_poor_high_low=true;
  show_initial_balance=false;
  show_session_header=true;
  highlight_start_end=true;
  profile_spacing=1;

  has_data(symbol:string){ return this.data.has(symbol) && this.data.get(symbol)!.length>0; }
  get_sessions(symbol:string){ return this.data.get(symbol) ?? null; }
  toggle_expand(symbol:string, idx:number){ const s=this.data.get(symbol); if (s && idx>=0 && idx<s.length) s[idx].expanded=!s[idx].expanded; }
  clear(symbol:string){ this.data.delete(symbol); this.last_build_hash.delete(symbol); }
  clear_all(){ this.data.clear(); this.last_build_hash.clear(); }

  build_sessions(symbol:string, timestamps:number[], highs:number[], lows:number[], candle_count:number, timeframe_sec:number, tick_per_row:number){
    if (!timestamps || !highs || !lows || !candle_count){ this.data.set(symbol, []); return; }
    const hash = (timestamps.slice(-10).join(',')+highs.slice(-10).join(',')+lows.slice(-10).join(',')+timeframe_sec+tick_per_row).length;
    if (this.last_build_hash.get(symbol)===hash && this.data.has(symbol)) return;
    this.last_build_hash.set(symbol, hash);
    const session_ms=this.session_period_hours*3600*1000;
    const sessions_dict=new Map<number, number[]>();
    for (let idx=0; idx<candle_count; idx++){
      const ts=timestamps[idx];
      const sess_start=Math.floor(ts/session_ms)*session_ms;
      if (!sessions_dict.has(sess_start)) sessions_dict.set(sess_start, []);
      sessions_dict.get(sess_start)!.push(idx);
    }
    const sessions:TPOSession[]=[];
    for (const sess_start of Array.from(sessions_dict.keys()).sort((a,b)=>a-b)){
      const indices=sessions_dict.get(sess_start)!;
      if (!indices.length) continue;
      const sess_end=sess_start+session_ms;
      const sess_highs=indices.map(i=>highs[i]);
      const sess_lows=indices.map(i=>lows[i]);
      const s_high=Math.max(...sess_highs);
      const s_low=Math.min(...sess_lows);
      if (s_high<=s_low) continue;
      let tpr=tick_per_row;
      if (tpr<=0){
        let prange=s_high-s_low;
        tpr=prange/40;
        if (tpr>0){
          const mag=Math.pow(10, Math.floor(Math.log10(tpr)));
          tpr=Math.round(tpr/mag)*mag;
          if (tpr===0) tpr=mag;
        } else tpr=1;
      }
      let row_count=Math.max(1, Math.ceil((s_high-s_low)/tpr));
      if (row_count>500){ tpr=(s_high-s_low)/500; row_count=500; }
      const rows:TPORow[]=[];
      for (let r_idx=0;r_idx<row_count;r_idx++){
        const lo=s_low+r_idx*tpr;
        const hi=lo+tpr;
        rows.push({price_lo:lo, price_hi:hi, row_idx:r_idx, blocks:[], block_count:0, is_poc:false, is_value_area:false, is_single_print:false, is_initial_balance:false});
      }
      for (let period_idx=0; period_idx<indices.length; period_idx++){
        const c_idx=indices[period_idx];
        const ch=highs[c_idx], cl=lows[c_idx];
        const start_row=Math.max(0, Math.floor((cl-s_low)/tpr));
        const end_row=Math.min(row_count-1, Math.floor((ch-s_low)/tpr));
        for (let r=start_row;r<=end_row;r++){ rows[r].blocks.push({period_idx}); rows[r].block_count=rows[r].blocks.length; }
      }
      const sess:TPOSession={
        session_start_ms:sess_start, session_end_ms:sess_end, rows, total_periods:indices.length, total_blocks:0, max_block_count:0,
        poc_price:0, poc_row_idx:-1, vah:0, val:0, session_high:s_high, session_low:s_low, ib_high:0, ib_low:0,
        has_poor_high:false, has_poor_low:false, tick_per_row:tpr, expanded:false
      };
      if (rows.length){
        let max_row=rows[0];
        for (const r of rows) if (r.block_count>max_row.block_count) max_row=r;
        sess.poc_price=(max_row.price_lo+max_row.price_hi)/2;
        sess.poc_row_idx=max_row.row_idx;
        max_row.is_poc=true;
        sess.max_block_count=max_row.block_count;
        sess.total_blocks=rows.reduce((s,r)=>s+r.block_count,0);
        const target=sess.total_blocks*this.value_area_pct;
        const sorted=Array.from(rows).sort((a,b)=>b.block_count-a.block_count);
        let va_vol=0; const va_prices:number[]=[];
        for (const r of sorted){ va_vol+=r.block_count; va_prices.push((r.price_lo+r.price_hi)/2); if (va_vol>=target) break; }
        if (va_prices.length){ sess.vah=Math.max(...va_prices); sess.val=Math.min(...va_prices); for (const r of rows){ const mid=(r.price_lo+r.price_hi)/2; if (mid>=sess.val && mid<=sess.vah) r.is_value_area=true; } }
        for (const r of rows) if (r.block_count===1) r.is_single_print=true;
        if (rows[0].block_count>=2) sess.has_poor_low=true;
        if (rows[rows.length-1].block_count>=2) sess.has_poor_high=true;
        if (indices.length>=2){
          const ib_high=Math.max(highs[indices[0]], highs[indices[1]]);
          const ib_low=Math.min(lows[indices[0]], lows[indices[1]]);
          sess.ib_high=ib_high; sess.ib_low=ib_low;
          const ib_start=Math.max(0, Math.floor((ib_low-s_low)/tpr));
          const ib_end=Math.min(row_count-1, Math.floor((ib_high-s_low)/tpr));
          for (let r=ib_start;r<=ib_end;r++) rows[r].is_initial_balance=true;
        }
      }
      sessions.push(sess);
    }
    this.data.set(symbol, sessions);
  }
}
